### PR TITLE
Background monitor

### PR DESCRIPTION
**Pull request by matthiasclasen**
This is an implementation of the platform-dependent parts of https://github.com/flatpak/xdg-desktop-portal/pull/319

### CODE REVIEW

**Comment by matthiasclasen** [comment]
One issue with this: we are currently getting the wrong information from gnome-shell. It puts desktop filenames into the windows list, which we misinterpret as app ids.

**Comment by matthiasclasen** [comment]
gnome-shell issue: https://gitlab.gnome.org/GNOME/gnome-shell/issues/1289

File: src/lockdown.c [hunk 1]
```diff
@@ -62,7 +62,7 @@ lockdown_init (GDBusConnection *bus,
   if (g_settings_schema_has_key (schema, "disable-microphone"))
     g_settings_bind (privacy, "disable-microphone", helper, "disable-sound-output", G_SETTINGS_BIND_DEFAULT);
   if (g_settings_schema_has_key (schema, "disable-sound-output"))
-    g_settings_bind (privacy, "disable-sound-output", helper, "disable-sound", G_SETTINGS_BIND_DEFAULT);
+    g_settings_bind (privacy, "disable-sound-output", helper, "disable-sound-output", G_SETTINGS_BIND_DEFAULT);
```
Comment (hadess) [inline src/lockdown.c]: Should probably be in another PR.

File: src/background.c [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+          g_variant_lookup (dict, "app-id", "&s", &app_id);
+          g_variant_lookup (dict, "is-hidden", "b", &hidden);
+          g_variant_lookup (dict, "has-focus", "b", &focus);
+
+          if (app_id == NULL)
+            continue;
+
+          app = g_strdup (app_id);
+          if (g_str_has_suffix (app, ".desktop"))
+            app[strlen (app) - strlen (".desktop")] = '\0';
```
Comment (alexlarsson) [inline src/background.c]: This here is problematic. Flatpaks can validly have an app id that ends with .desktop. And in fact, `org.telegram.desktop` does this, which has triggered bugs in a bunch of places like this..

Comment (hadess) [reply src/background.c]: I think it's a hack until the gnome-shell bug mentioned in https://github.com/flatpak/xdg-desktop-portal-gtk/pull/198#issuecomment-492026996 is fixed.

File: src/lockdown.c (hunk 1 above)
Comment (matthiasclasen) [reply src/lockdown.c]: good point

File: src/background.c (hunk 2 above)
Comment (matthiasclasen) [reply src/background.c]: Yeah, sorry, should have commented on that. It is indeed a temporary hack

File: src/background.c [hunk 3]
```diff
@@ -262,6 +262,122 @@ handle_notify_background (XdpImplBackground *object,
...
+
+  if (g_mkdir_with_parents (dir, 0755) != 0)
+    {
+      g_warning ("Failed to create dirs %s", dir);
+      goto out;
+    }
+
+  args = g_string_new ("");
+  g_string_append (args, "flatpak run");
+  if (arg_commandline && arg_commandline[0])
```
Comment (alexlarsson) [inline src/background.c]: I think xdg-desktop-portal should generate the full commandline, because if the calling app is not a flatpak (like, maybe its not contained at all) it should not generate a flatpak run command, and the backend implementation should not have to care about this.

File: src/background.c [hunk 4]
```diff
@@ -262,6 +262,122 @@ handle_notify_background (XdpImplBackground *object,
...
+  g_autofree char *file = NULL;
+  g_autofree char *path = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GString) args = NULL;
+  g_autoptr(GKeyFile) keyfile = NULL;
+
+  g_debug ("background: handle EnableAutostart");
+
+  file = g_strconcat (arg_app_id, ".desktop", NULL);
+  dir = g_build_filename (g_get_user_config_dir (), "autostart", NULL);
```
Comment (alexlarsson) [inline src/background.c]: Shouldn't there be a "gnome" in this somewhere?

File: src/background.c [hunk 5]
```diff
@@ -262,6 +262,122 @@ handle_notify_background (XdpImplBackground *object,
...
+  g_autofree char *dir = NULL;
+  g_autofree char *file = NULL;
+  g_autofree char *path = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GString) args = NULL;
+  g_autoptr(GKeyFile) keyfile = NULL;
+
+  g_debug ("background: handle EnableAutostart");
+
+  file = g_strconcat (arg_app_id, ".desktop", NULL);
```
Comment (alexlarsson) [inline src/background.c]: Do we ever want callers to be able to generate multiple autostart files? I.e. do we want to pass in some kind of suffix for the filename?

File: src/background.c (hunk 4 above)
Comment (matthiasclasen) [reply src/background.c]: I checked the gnome-session code. XDG_CONFIG_HOME/autostart is where it looks.

File: src/background.c (hunk 5 above)
Comment (matthiasclasen) [reply src/background.c]: I don't see a need

File: src/background.c [hunk 6]
```diff
@@ -262,6 +262,83 @@ handle_notify_background (XdpImplBackground *object,
...
+      goto out;
+    }
+
+  if (g_mkdir_with_parents (dir, 0755) != 0)
+    {
+      g_warning ("Failed to create dirs %s", dir);
+      goto out;
+    }
+
+  commandline = g_strjoinv (" ", (char **)arg_commandline);
```
Comment (alexlarsson) [inline src/background.c]: I think this is where we should do any required escaping, rather than in the portal.

File: src/background.c [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
```
Comment (alexlarsson) [inline src/background.c]: You still need to escape here
